infrastructure/src/main/java/org/corfudb/infrastructure/management/ClusterStateCollector.java

```diff
@@ -0,0 +1,118 @@
+package org.corfudb.infrastructure.management;
+
+import com.google.common.collect.ImmutableMap;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.management.ClusterStateContext.HeartbeatCounter;
+import org.corfudb.protocols.wireprotocol.ClusterState;
+import org.corfudb.protocols.wireprotocol.NodeState;
+import org.corfudb.protocols.wireprotocol.SequencerMetrics;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeConnectivity;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeConnectivity.ConnectionStatus;
+import org.corfudb.runtime.exceptions.WrongEpochException;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+@Builder
+@Slf4j
+public class ClusterStateCollector {
+    @NonNull
+    private final String localEndpoint;
+    @NonNull
+    private final Map<String, CompletableFuture<NodeState>> clusterState;
+
+    private final long timeout;
+
+    @NonNull
+    private final HeartbeatCounter heartbeatCounter;
+
+    public ClusterState collectClusterState(long epoch, SequencerMetrics sequencerMetrics) {
+        Map<String, NodeState> nodeStates = new HashMap<>();
+
+        nodeStates.put(localEndpoint, collectLocalNodeState(epoch, sequencerMetrics));
+        nodeStates.putAll(collectNodeStates());
+
+        return ClusterState.builder()
+                .localEndpoint(localEndpoint)
+                .nodes(ImmutableMap.copyOf(nodeStates))
+                .build();
+    }
+
+    public ImmutableMap<String, Long> collectWrongEpochs() {
+        Map<String, Long> wrongEpochs = new HashMap<>();
+
+        clusterState.forEach((server, state) -> {
+            try {
+                //Add a node to connectedNodes list if node responded with its NodeState
+                state.get(timeout, TimeUnit.MILLISECONDS);
```

Comment:
Please remove this duplicate code and aggregate all lists in a single place. It gets difficult to understand the code.

There can also be the following scenario:
Since we have an RPC timeout of 5 seconds.
In this code, every method will timeout differently:
First, you would collect node states and timeout in 2 seconds.
Then you collect wrong epoch exceptions where your future could have completed and returned a wrongEpochException.

Reply:
Actually, it gets simpler. Let me explain:
1. Single responsibility principle - why that one is one of the most important ones? This principle makes code much easier to understand. In the previous version of this algorithm, we had one `for cycle` where we update **3** collections, which brings next problems:
    - it increases cognitive complexity (it gets much harder to understand)
    - makes code error prone
    - prevents us from writing unit tests! This part is most important. The unit tests they are indicators, if you can write a unit test easily for your code, that means your code has good quality. You can see I covered ClusterStatusCollector by unit tests easily.

Now when all 3 parts of the `collecting` algorithm are independent, it's easier to understand, it's compostable, all parts are independent and covered by unit tests. 
There is a little bit more code, but it's worth it.

Reply:
Regarding completable futures, you don't pay any extra cost if you navigated many times. 
Let me explain this part: when the completable future gets resolved you can get the result from the future immediately (you don't excecute CF many times, you execute it ones and then you reuse results as many times as you want)

Reply:
I don't feel this code is simplifying the current code. It makes it harder to read.
Another thing,
Going back to my previous observation, this is wrong because of the scenario I mentioned:
you wait on the same completable future twice for 2 seconds each. The completable future may not complete in the first 2 seconds but might complete in the 3rd second and return you a valid value.

Reply:
Regarding timeouts you are right, it could happen in theory. 
Also, since we set up a client router to resolve the future after some timeout, we don't need to have the same timeout in the collector.  ClusterCollector shouldn't be in charge of timeouts. 
I have deleted all timeouts from the collector.

Reply:
Regarding simplicity - if you can prove somehow that prev logic was better, I would agree. For now, on my side, I have objective arguments: best practices, and unit tests.